[WIP] Transaction tests using simulated backend

sebastiandelgado opened the pull request:
Closes #812 

Done:
 - Single test for all simulated backend test cases implementing ethTransfer and contractCreation test cases

Missing:
 - Test case for contract state transition needs to check that the contract has changed owner in the validation function

dshulyak commented on geth/transactions/transactor_test.go:
```diff
@@ -385,41 +386,72 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
+	}{
+		{
+			"EthTransfer",
+			hexutil.Bytes(gethcommon.FromHex("0.1")),
```
i dont think that eth transfer has any input in tx

sebastiandelgado replied:
My bad! Needed that in the .Value field. Pushing fix now

sebastiandelgado commented:
@dshulyak Thanks for the pointers today. I think the tests are almost done. I'm running into the following error in the state transition test case:
```
ERROR[06-15|13:59:56] queued transaction does not belong to the selected account package=status-go/transactions.Manager err="transaction can only be completed by its creator"
```
I'm using the same address for the selected account and the tx.From field in the test. When I log them, they show as being the same, but for some reason the completeFunc in the transactor gets a selected account address thats different from the one I'm setting in the test. Any idea why this may be happening?

dshulyak commented:
> When I log them, they show as being the same, but for some reason the completeFunc in the transactor gets a selected account address thats different from the one I'm setting in the test. Any idea why this may be happening?

tbh i don't know. it should be possible to trace the execution and find out why it happens

sebastiandelgado commented on transactions/transactor_test.go:
```diff
@@ -385,41 +388,126 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
+				}
+			},
+			func(receipt *types.Receipt, testaddr common.Address) {
+				successStatus := uint(1)
```
@dshulyak This PR should be ready. The only thing missing is getting the contract owner in the response to the last transaction in this last test case. Currently we are only checking that there is a success status in the transaction receipt, but I need to figure out how to get the response to the Owner method in the contract in order to double check that there was an owner change for that particular ENS node

sebastiandelgado replied:
Still need to figure out how to do this, but it turns out the tests were passing because they were cached. Will fix them and take off the WIP tag again when its done

sebastiandelgado commented:
Also looking into the `make test-unit-race` errors

sebastiandelgado commented:
This is working except that I can't figure out how to resolve a race condition between `TestLocalNonce` and `TestTransactionsWithSimulatedBackend` thats coming up with ` go test ./transactions/ -race`. I'm still looking at what it is between these two tests that interfere with each other that isn't interfering with the other tests. It seems they are trying to approve each others requests from the `s.manager.pendingSignRequests` and failing because of different selectedAccounts. Will update on any further progress/problems here.

dshulyak commented on transactions/transactor_test.go:
```diff
@@ -313,6 +316,132 @@ func (s *TxQueueTestSuite) TestDiscardTransaction() {
+
+			go func() {
+				transactionsApproved := 0
+				for i := 1000; i > 0; i-- {
```
this loop doesn't make any sense. i see that you didnt add it. but i it is weird and wrong, why to approve 1000 txs? and this is a source of those warnings about wrong key for approval

dshulyak commented:
the problem with those approval loops in both tests. i think that is a simply wrong pattern and should be reworked. i experimented a bit with waiting till that approval loop is closed

```
 // TestLocalNonce verifies that local nonce will be used unless
 // upstream nonce is updated and higher than a local
 // in test we will run 3 transaction with nonce zero returned by upstream
@@ -328,10 +460,21 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
 		AccountKey: &keystore.Key{PrivateKey: key},
 	}
 	nonce := hexutil.Uint64(0)
-
+	waiter := make(chan struct{})
+	defer func() {
+		<-waiter
+	}()
+	closer := make(chan struct{})
+	defer close(closer)
 	go func() {
 		approved := 0
 		for {
+			select {
+			case <-closer:
+				close(waiter)
+				return
+			default:
+			}
 			// 3 in a cycle, then 2
 			if approved >= txCount+2 {
 				return
@@ -384,42 +527,3 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
 	resultNonce, _ = s.manager.localNonce.Load(args.From)
 	s.Equal(uint64(nonce)+1, resultNonce.(uint64))
 }
```

sebastiandelgado commented:
@dshulyak Thanks for the pointers on using channels to contain each test. I applied them to the nonce test as you suggested and to the simulated backend tests as well. 

The last thing I think I need to fix here is that I'd like to check the result of the transactions to verify that the new owner has been set on the ens node. I made some comments on how that would be done, but for some reason I'm never getting back any response from the Wait() method. I added comments on the files for you to see exactly what I mean.

sebastiandelgado commented on transactions/transactor_test.go:
```diff
@@ -385,41 +401,152 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
+						numTransactions++
+						if numTransactions == len(transactions) {
+							go func() {
+								getFinalResult <- s.manager.pendingSignRequests.Wait(req.ID, time.Second*1)
```
Here is where I'm trying to get the result from the sign request in order to read it below and pass it to the 'receiptValidator' function, which I should probably also rename. 

This simply is not returning at the moment. Do you know if the simulated backend even allows for that?

Commit:
Simulated backend tests for contract creation, eth transfer and contract state transfer

dshulyak commented on transactions/transactor_test.go:
```diff
@@ -385,41 +401,152 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
+				// Give the node a value so that we are reassigning ownership of something
+				// other than the 0 node, which already belongs to the current owner
+				node[0] = []byte("1")[0]
+				ensParsed, _ := abi.JSON(strings.NewReader(contract.ENSABI))
```
error can be checked

dshulyak commented on transactions/transactor_test.go:
```diff
@@ -385,41 +401,152 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
+				node[0] = []byte("1")[0]
+				ensParsed, _ := abi.JSON(strings.NewReader(contract.ENSABI))
+				secondOwnerAddress := account.FromAddress(TestConfig.Account1.Address)
+				setOwnerMethodData, _ := ensParsed.Pack("setOwner", setOwnerMethodID, node, secondOwnerAddress)
```
why do you pack method id?
the result of pack should be 4 bytes of method hash and packed inputs. in this case node and second owner

sebastiandelgado replied:
I believe that method name is the first arg of the pack function and then the first argument is the 4byte method hash. Going off of this comment: https://github.com/status-im/status-go/blob/develop/vendor/github.com/ethereum/go-ethereum/accounts/abi/abi.go#L47

dshulyak commented on transactions/transactor_test.go:
```diff
@@ -385,41 +401,152 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
+
+				// method ID: web3.utils.sha3('owner(bytes32)').substring(0, 8)
+				ownerMethodID := "0x02571b"
+				getOwnerMethodData, _ := ensParsed.Pack("owner", ownerMethodID, node)
```
same, method id already packed

dshulyak commented on transactions/transactor_test.go:
```diff
@@ -385,41 +401,152 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
+			func(receipt *types.Receipt, finalResult sign.Result, testaddr common.Address) {
+				successStatus := uint(1)
+				s.Equal(successStatus, receipt.Status)
+				// secondOwnerAddress := account.FromAddress(TestConfig.Account1.Address)
```
please dont leave commented code in source

dshulyak commented on transactions/transactor_test.go:
```diff
@@ -385,41 +401,152 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
+				<-waiter
+			}()
+			closer := make(chan struct{})
+			defer close(closer)
```
i know that i proposed this solution. but it can be implemented in a re-usable way

dshulyak commented on transactions/transactor_test.go:
```diff
@@ -385,41 +401,152 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
-	tx := SendTxArgs{
-		From:  testaddr,
-		Input: hexutil.Bytes(gethcommon.FromHex(contract.ENSBin)),
+func (s *TxQueueTestSuite) TestTransactionsWithSimulatedBackend() {
```
to be honest it is very hard to read table driven tests written this way. i think it is just better to introduce couple of helpers that re-used accross different normally written tests

dshulyak commented on transactions/transactor_test.go:
```diff
@@ -385,41 +401,152 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
+			}()
+
+			// The receipt from the last transaction in the testcase
+			var finalReceipt *types.Receipt
```
i would check every receipt for every transaction and changed state associated with that transaction

dshulyak commented on transactions/transactor_test.go:
```diff
@@ -385,41 +401,152 @@ func (s *TxQueueTestSuite) TestLocalNonce() {
+				node := make([]byte, 32)
+				// Give the node a value so that we are reassigning ownership of something
+				// other than the 0 node, which already belongs to the current owner
+				node[0] = []byte("1")[0]
```
you can also do [32]byte{1} , which will serve the purpose. or if you want to use slice
for some reason you still can just set any element without conversions

Commit:
Split test cases up into separate tests with helpers

sebastiandelgado commented:
Added [WIP] tag again to address Dmitry's comments.

Done:

 - Split up large test with testcase to 3 smaller tests with helpers to make them easier to understand
 - Check each receipt for success status

Working on:
 - Packing the correct method data into the contract state transfer test transactions + checking for errors on ensParsed.Pack return value
 - Get and check the result of each transaction
 - Check the updated state after transferring eth
 - Implement the waiter/closer pattern in a reusable way

mandrigin reviewed (approved):
lgtm in terms of smart contract calls

Commit:
Implement the waiter/closer pattern in a reusable way

sebastiandelgado commented:
Just implemented the waiter/closer pattern in a reusable way. Still missing:

 - Packing the correct method data into the contract state transfer test transactions + checking for errors on ensParsed.Pack return value
 - Get and check the result of each transaction
 - Check the updated state after transferring eth

dshulyak commented on transactions/transactor_test.go:
```diff
@@ -396,30 +437,123 @@ func (s *TxQueueTestSuite) TestContractCreation() {
+		},
+		{
+			// Check that the contract owner is correct
+			Input: hexutil.Bytes(getOwnerMethodData),
```
getOwner is not a transaction but a call that reads from storage. it cant have a receipt

sebastiandelgado commented:
Pausing work on this as I work on #1070, since that one will significantly change the way these tests are written.
